### PR TITLE
Add pre-build Docker images through GHCR

### PR DESCRIPTION
**Pull request by JJ-8**
As discussed in https://github.com/TFNS/CTFNote/discussions/211, I have created the following:

- The workflow builds and pushes the front, api, db images to GHCR
- The images are both for amd64 as well as arm64
- The images are properly tagged with `latest`, by the branch name and by tag
    - Please note that I was not able to test by tag, but I am pretty sure this works fine since the other two also work as expected
- The front, api and db images are build concurrently using a matrix build for maximum speed
    - Still the build of the front image takes around 20 minutes due to the arm build
- A `docker-compose.prebuild.yml` is provided which extends the original `docker-compose.yml` to use the pre-build images from GHCR. This gives the option to the user to build the images themself too.
- Instructions are added to the README
- A fix was added for building the front Docker container

It is not possible to test this PR before merging. However, you can look at [the packages of my fork](https://github.com/JJ-8/CTFNote/packages) to see the results. Please note that you should enable the package view on the homepage manually in GitHub through the settings icon (also used to change the description of the project).

For now it is only GHCR because that is the easiest to implement and does not require you to setup secrets. In the near future (as soon as we have fully decided which external Docker hosting platform we are going to use), it is almost trivial to add this as a target too due to the `docker/metadata-action` step.

### CODE REVIEW

**Comment by JJ-8** [comment]
@XeR, would it be possible to revert to merge commits again for merging the PRs? Fast forwarding messes up the commits in the fork (see above commits which are getting marked as new but do not add any content). Furthermore, the [network overview in GitHub](https://github.com/TFNS/CTFNote/network) is wrong since there is no explicit merge done (thus not showing my branches merging into the upstream branch). Unless you highly prefer fast forwarding, merge commits would make it easier for me to contribute to the project :)

File: front/Dockerfile [hunk 1]
```diff
@@ -1,5 +1,6 @@
 FROM node@sha256:1a9a71ea86aad332aa7740316d4111ee1bd4e890df47d3b5eff3e5bded3b3d10 AS build-deps
 
+RUN apk add --no-cache python3 make g++
```
Comment (XeR) [inline front/Dockerfile:3]: Why is this required ? Is this arm64 only ?
What requires Python here ?

I could track this down to https://github.com/anthares101/CTFNote/commit/2aff33507ff52816a44e114271d54480c133eec1 but I am not sure what is going on.
Is this related to a node update ? If yes, does this apply to us ?

File: README.md [hunk 2]
```diff
@@ -12,13 +12,31 @@ CTFNote is a collaborative tool aiming to help CTF teams to organise their work.
...
 Before starting, make sure to fill in the information in the `.env` file.
 
-Then you can start it with `docker-compose`. The default
+### Pre-build images
+
+Building CTFNote requires at least 3 GB of RAM. If you want to host CTFNote
+on a server with less than 3 GB of RAM, you can use the pre-build images
+from the GitHub Container Registry.
+
+Make sure to [authenticate Docker to GHCR](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#authenticating-with-a-personal-access-token-classic).
```
Comment (XeR) [inline README.md:21]: Is this really required ? Isn't this only for pushing and pulling from private repos ?
I could download your packages without providing a token.

Command: `docker image pull ghcr.io/jj-8/ctfnote/db`
Maybe I logged in a long time ago and don't remember it ?

File: .github/workflows/build-and-publish.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        runs-on: ubuntu-latest
+        strategy:
+            fail-fast: false
+            matrix:
+                component: ["front", "api", "db"]
+        steps:
+            - name: PrepareReg Names
+              run: |
+                  echo IMAGE_REPOSITORY=$(echo ${{ github.repository }} | tr '[:upper:]' '[:lower:]') >> $GITHUB_ENV
+                  echo IMAGE_TAG=$(echo ${{ github.ref }} | tr '[:upper:]' '[:lower:]' | awk '{split($0,a,"/"); print a[3]}') >> $GITHUB_ENV
```
Comment (XeR) [inline .github/workflows/build-and-publish.yml:27]: I wonder if we can inject a command in `github.ref`. If we can, what are the implications ? Can we make a challenge out of this ? :-)

**Review by XeR — CHANGES_REQUESTED** [review]
I'm trusting you blindly on this new yaml file. If it works for your repo, it'll probably work here too

**Comment by XeR** [comment]
> @XeR, would it be possible to revert to merge commits again for merging the PRs? Fast forwarding messes up the commits in the fork (see above commits which are getting marked as new but do not add any content). Furthermore, the [network overview in GitHub](https://github.com/TFNS/CTFNote/network) is wrong since there is no explicit merge done (thus not showing my branches merging into the upstream branch). Unless you highly prefer fast forwarding, merge commits would make it easier for me to contribute to the project :)

Sure !
I should have read this before merging other PR, sorry.

File: front/Dockerfile (hunk 1 above)
Comment (JJ-8) [reply front/Dockerfile:3]: It is quite a long time ago, but I remember that this was related to node-gyp on arm64 builds of CTFNote. These tools should be installed in order to build CTFNote successfully.

File: README.md (hunk 2 above)
Comment (JJ-8) [reply README.md:21]: I remember that I had authorization issues in the beginning with GitHub when pulling the images. I also saw [this](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#authenticating-to-the-container-registry) in the documentation:
> You need an access token to publish, install, and delete private, internal, and public packages.

But if it is actually not required, we can drop the requirement of course.

File: .github/workflows/build-and-publish.yml (hunk 3 above)
Comment (JJ-8) [reply .github/workflows/build-and-publish.yml:27]: If you change your name to some command injection payload, or someone forks CTFNote and changes the repository name to some command injection payload, I think it is definitely possible to attack yourself through your own repository (but not the other forks or upstream)! ;-)

File: front/Dockerfile (hunk 1 above)
Comment (XeR) [reply front/Dockerfile:3]: Ah, thanks.
Now that I think about it, I'm surprised this works though : the images are pinned to amd64 versions. How does that even work ? Is Docker flat out ignoring digests ?

File: README.md (hunk 2 above)
Comment (XeR) [reply README.md:21]: I am quite confused. The document you quoted appears to say that a token is required, but I can download image without being authenticated.
Your tokens are stored in `~/.docker/config.json`. Can you try to download your image with no tokens ?
If you can confirm that pulling an image does not require a token, please remove this part from the documentation

File: front/Dockerfile (hunk 1 above)
Comment (JJ-8) [reply front/Dockerfile:3]: I do not understand your question unfortunately. I think these build tools are probably required by `yarn` in order to compile certain native dependencies into another architecture. Therefore I think it worked without arm64 but now these are required since we include arm64.

Comment (XeR) [reply front/Dockerfile:3]: Ah, sorry for the confusion.

My question refers to this line in particular :
```
FROM node@sha256:1a9a71ea86aad332aa7740316d4111ee1bd4e890df47d3b5eff3e5bded3b3d10
```

This container is pinned to a specific digest.

If we try to use a digest that is not for our architecture, Podman warns us about that :
```
Trying to pull docker.io/library/node@sha256:bd59aa24c8e6d84170b0f4f17e8b94ba21ee6a70b989e81bb793af93dad351a3...
[...]
WARNING: image platform (linux/s390x) does not match the expected platform (linux/amd64)
```

And when I try to create a container from the newly-pulled image, both Docker and Podman are panicking :
```
WARNING: The requested image's platform (linux/s390x) does not match the detected host platform (linux/amd64/v3) and no specific platform was requested
exec /usr/local/bin/docker-entrypoint.sh: exec format error
```

So my question is : why does this work ?

Comment (XeR) [reply front/Dockerfile:3]: I'm fine with merging this if we remove everything related to arm64 and postpone new archs to a different PR.
I believe we'd have to remove this changes to this file, change the platforms from the yaml file and maybe remove the qemu step (but I'd rather not touch this file, since I have no idea how to test all of this)

Comment (anthares101) [reply front/Dockerfile:3]: Hello, I can verify that the arm64 built is working as expected. I actually created the fork for that, the thing is that you are right, it should not work but it does 🤣. I believe that it is related to Qemu, the Github actions runner is amd64 after all and maybe Qemu perform the necessary steps to fix the problems to cross compile the image.

Comment (XeR) [reply front/Dockerfile:3]: Hey anthares101, thanks for your work. Your feedback on this is greatly appreciated.

It makes absolutely no sense, yet it works.

Assuming you have static qemu binaries installed and binfmt configured properly, you can reproduce the build with (this is slow):
```sh
docker buildx build --platform linux/arm64 front
```

I am not sure if I should be impressed that there is some logic to find a similar version of a pinned container for an other architecture or scared that Docker ignores digests whenever it feels like.

**Review by XeR — APPROVED** [review]
